rfd/0101-pod-rbac.md

```diff
@@ -0,0 +1,825 @@
+API for the following two endpoints:
+
+- `/api/v1/pods`
+  - `GET`: list or watch objects of kind Pods in every namespace.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `GET`: list or watch objects of kind Pods in namespace `{namespace}`.
+
+Once Teleport receives the response from Kubernetes API, it will unmarshal the
+response and exclude any Pod that is not in the allowed list. After excluding them,
+Teleport will return the filtered response to the user.
+
+##### Deleting Pods
+
+Kubernetes allows a user to delete every Pod in a certain namespace using a single
+call. The call does not include the name of the Pods to delete, so Teleport won't
+forward the request to Kubernetes. Instead, Teleport will list the user's Pods,
+exclude the ones that he shouldn't have access to, and issue a delete request for
+each one of the resulting Pod names.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `DELETE`: delete a collection of Pods in namespace `{namespace}`.
+
+Another option is to prevent this type of request in Teleport. This means that
+if you try to delete every Pod in the namespace while your role limits the Pods
+you have access to, Teleport will reject the request and inform the user that
+he is not allowed to do it. This option will restrict one of the `kubectl`
+functionalities.
+
+##### Creating Pods
+
+The endpoint for creating pods won't be restricted, and if restrictions exist
+it's the Kubernetes RBAC layer that applies them.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `POST`: create a POD in namespace `{namespace}`.
+
+### Security
+
+When a user has multiple roles assigned and his access is expected to be restricted
+to a subset of the pods available, it's a requirement that the union of every
+`kubernetes_resources` defined does not give broader permissions as exemplified by `user4` in
+[Restricted Pod Access](#restrict-pod-access-selected) examples. Otherwise,
+Teleport will leak the Pods when the user lists them.
+
+When a user lists pods - `kubectl get pods`, using the Kubernetes RBAC given by
+`search_as_roles` or by a Kubernetes RBAC principal that allows listing pods
+before requesting access to a certain Pod, the user has access
+to other details besides the Pod name. It happens because `kubectl get pods -o json|yaml`
+API response returns Pod description that includes env variables, secrets used,
+and other details. This might be a security concern because Pods YAML might contain
+secrets that can be leaked even before creating an access request.
+
+When a user has the ability to `exec` into a Pod, it's also a security concern
+because at that time Teleport access control can be bypassed and the user is able
+to send requests directly to the Kubernetes API endpoint using the Pod's credentials.
+If the target Pod has RBAC permissions atached such as `Impersonation`, delete pods, logs...
+the user bypasses Teleport and is able to execute the commands. Take into consideration
+each RBAC policy attached to each pod.
```

Comment:
Similarly to the point above: is it possible to at least detect these cases and warn admins that Teleport RBAC can be bypassed in these cases?